### PR TITLE
ci: Github Actions workflows for publish to marketplace

### PR DESCRIPTION
**Pull request by gbockus-sf**
### What does this PR do?
- Removed the circleci config file
- Removed all refs to circleci in the docs & code. 
- Add new Github action workflows for handling publishing from github actions
- Doc'd required GHA setup in quip. [link](https://salesforce.quip.com/6OCzAghbtCdC)
- Update the build & test workflow to be more reusable
- Add a new slack job for general posts to slack. 

### What issues does this PR fix or reference?
@W-11596486@
@W-11871656@

### Functionality Before
publishes happen in circleCI

### Functionality After
publishes happen in github actions

### CODE REVIEW

File: .github/workflows/buildAndTest.yml
Comment (gbockus-sf) [inline .github/workflows/buildAndTest.yml:1]: Renamed since we use this on nightly build & commit workflow & publish workflow.

File: .github/workflows/mergeReleaseBranch.yml
Comment (gbockus-sf) [inline .github/workflows/mergeReleaseBranch.yml:1]: This workflow will read the version in main and the provided release branch.  As long as the release branch is `newer` then main it will allow the merge into main. 

The merge into main will trigger the publish workflow to start https://github.com/forcedotcom/salesforcedx-vscode/pull/4606/files#diff-ff0df3629970528f0ac692a38deabc071dc87fdb0112519bba29a20a1023a028

File: .github/workflows/publish.yml
Comment (gbockus-sf) [inline .github/workflows/publish.yml:21]: This environment will have an approval in the github action configuration.  This is how we pause for github actions. 

Note the flow is 
Tests, Build All  -> publish -> slack notification -> wait for approval. 

Where we use to have two approvals for a publish we now have one. The only reason I kept that approval is it seems like a good step to install and verify the built vsix files before shipping it to microsoft

File: .github/workflows/publish.yml
Comment (gbockus-sf) [inline .github/workflows/publish.yml:59]: So I think we'll be able to use these ctc related workflows as is...we'll need to set a SVC-cli token in our repo but I think it'll be fine.

File: .github/workflows/publish.yml
Comment (gbockus-sf) [inline .github/workflows/publish.yml:71]: Note this is using the vscode-publish.result to determine the ctc result.  This output is set based on if we published to the vscode marketplace or not.  I assume if we get into the market place we should call it a success from a ctc perspective.

File: .github/workflows/publishVSCode.yml
Comment (gbockus-sf) [inline .github/workflows/publishVSCode.yml:14]: note we have two outputs here...one to specify if we made it to the marketplace..the other to indicate a fully successful run.

File: .github/workflows/publishVSCode.yml
Comment (gbockus-sf) [inline .github/workflows/publishVSCode.yml:66]: This is different then the old flow.  The old flow would generate new vsix files then publish those files in each package directory.  Since we are downloading built artifacts from GHA I just make a 1 liner for doing the publish vsce call.

File: .github/workflows/publishVSCode.yml
Comment (gbockus-sf) [inline .github/workflows/publishVSCode.yml:90]: note these jobs all depend on the previous step completing.  My intention here is if a step fails due to a transient error we can rerun from failed and complete the publish steps without too much manual intervention.

File: .github/workflows/slackNotification.yml
Comment (gbockus-sf) [inline .github/workflows/slackNotification.yml:68]: new job for a general notification.

File: contributing/publishing.md
Comment (gbockus-sf) [inline contributing/publishing.md:101]: I outlined how you can download the vsix artifacts from GHA here and how to easily install those.  let me know if you think we need an in between step for something closer to ctv, but it seems pretty straightforward.

File: .github/workflows/mergeReleaseBranch.yml
Comment (randi274) [reply .github/workflows/mergeReleaseBranch.yml:1]: Nit: Capitalize Branch; the rest sounds good

Comment (randi274) [inline .github/workflows/mergeReleaseBranch.yml:1]:
```suggestion
name: 'Merge Release Branch'
```

File: contributing/publishing.md
Comment (randi274) [inline contributing/publishing.md:47]:
```suggestion
After all tests have passed it will trigger a required approval for continuing the job.
```

File: contributing/publishing.md
Comment (randi274) [inline contributing/publishing.md:81]:
```suggestion
The steps used to publish to the VS Code Marketplace can be found in the associated Github Actions.
```

File: contributing/publishing.md
Comment (randi274) [inline contributing/publishing.md:91]:
```suggestion
- Download directly from the Github Action run. You will find artifacts that are associated with a run at the bottom of the summary screen
```

**Review by randi274 — COMMENTED** [review]
A first pass at reading and thinking through all of it - this is gonna be so great to get up and running again!

File: .github/workflows/publish.yml
Comment (gbockus-sf) [reply .github/workflows/publish.yml:21]: agreed..def view this as a first pass and we can continue to improve as we go along

File: .github/workflows/publishVSCode.yml
Comment (gbockus-sf) [reply .github/workflows/publishVSCode.yml:14]: unset I believe.  This is all new territory so I'm not 100% clear on how it behaves.

File: .github/workflows/publishVSCode.yml
Comment (gbockus-sf) [reply .github/workflows/publishVSCode.yml:90]: yeah I thought about that as well but in the end figured it'd be simpler to get all this working then adjust if it makes sense.  It does seem like it would make sense to publish on a tag creation, but I didn't want to alter the current flow until we get a better feel for it.

File: .github/workflows/publish.yml
Comment (gbockus-sf) [reply .github/workflows/publish.yml:4]: Looks like there's a default 30day timeout, but we can set it to be whatever we want https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments#about-required-reviews-in-workflows

File: .github/workflows/prepublish.yml
Comment (gbockus-sf) [reply .github/workflows/prepublish.yml:10]: I'm having trouble wrapping my brain around this.  LIke this should be an output from the cut release branch job?   With an approval that pauses until we're ready to publish? Let's chat when you're back

File: .github/workflows/publish.yml
Comment (klewis-sfdc) [inline .github/workflows/publish.yml:15]: 👍🏽

File: .github/workflows/publishVSCode.yml
Comment (klewis-sfdc) [reply .github/workflows/publishVSCode.yml:18]: Just bumping this one to see if you want to commit these changes.

File: .github/workflows/publishVSCode.yml
Comment (klewis-sfdc) [reply .github/workflows/publishVSCode.yml:72]: Does this work either way (with or without the '')?

File: contributing/publishing.md
Comment (klewis-sfdc) [inline contributing/publishing.md:49]: 👍🏽

File: contributing/publishing.md
Comment (klewis-sfdc) [inline contributing/publishing.md:100]: 💯

**Review by klewis-sfdc — APPROVED** [review]
Looks good to me, Gordon.  Let's get this thing running!  Thanks for all of the work getting these set up!

**Pull request merged**
